jdaviz/configs/specviz/plugins/parsers.py

```diff
@@ -118,33 +117,82 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
                     app.add_data_to_viewer("spectrum-viewer", data_label[i])
 
         if isinstance(data, SpectrumList):
-            if data[0].meta.get("INSTRUME") == "MIRI":
-                # reset display ranges, or build combined spectrum,
-                # when input is a SpectrumList instance and the
-                # observations were produced by MIRI
-                wlallarr = np.array(wlallorig)
-                fnuallarr = np.array(fnuallorig)
-                dfnuallarr = np.array(dfnuallorig)
-                srtind = np.argsort(wlallarr)
-                wlall = wlallarr[srtind]
-                fnuall = fnuallarr[srtind]
-                fnuallerr = dfnuallarr[srtind]
-
-                # units are not being handled properly yet.
-                unc = StdDevUncertainty(fnuallerr * flux_units)
-                spec = Spectrum1D(flux=fnuall * flux_units, spectral_axis=wlall * wave_units,
-                                  uncertainty=unc)
-
-                # Make metadata layout conform with other viz.
-                spec.meta = standardize_metadata(spec.meta)
-
-                # needs perhaps a better way to label the combined spectrum
-                label = "Combined " + data_label[0]
-                app.add_data(spec, label)
-                if show_in_viewer:
-                    app.add_data_to_viewer("spectrum-viewer", label)
-            else:
-                # show only the first FITS extension if the observations
-                # were produced by instrument other than MIRI
-                if show_in_viewer:
-                    app.add_data_to_viewer("spectrum-viewer", data_label[0])
+            # If >1 spectra in the list were opened from the same FITS file,
+            # group them by their original FITS filenames and add their combined
+            # 1D spectrum to the DataCollection.
+            unique_files = group_spectra_by_filename(app.data_collection)
```

Comment:
This logic adds a composite spectrum to the data collection for every unique FITS file loaded via, for example, `SpectrumList.read`. This approach allows users to pass in heterogeneous `SpectrumList`s originating from multiple instruments, and will concatenate the spectra with multiple orders only when they originate from the same data product.

Reply:
I don't understand how this plays together with this line above:

https://github.com/spacetelescope/jdaviz/blob/eb9cfd5e6fddd690e10460f8ac5f90692bb548d3/jdaviz/configs/specviz/plugins/parsers.py#L96

You sure we're not duplicating the initially loaded stuff here?

Reply:
I'm not sure I follow. The `for i, spec in enumerate(data)` block is adding the individual spectra one at a time to the data collection and/or viewer. The `if concat_by_file and isinstance(data, SpectrumList)` block computes "combined"/"concatenated" spectra for the contents of each loaded FITS file, and adds the combined spectra to the collection/viewer. I don't think there's duplication here.

Reply:
Ah, gotcha. I got lost in the code, so I appreciate you made it clear to me.